theano/tensor/nnet/abstract_conv2d.py

```diff
@@ -0,0 +1,686 @@
+                                                        border_mode, subsample, filter_flip)
+
+    ## Update shape/height_width
+    def make_node(self, kern, topgrad, shape):
+        if kern.type.ndim != 4:
+            raise TypeError('kern must be 4D tensor')
+        if topgrad.type.ndim != 4:
+            raise TypeError('topgrad must be 4D tensor')
+        if self.subsample != (1, 1) and shape is None:
+            raise ValueError('shape must be given if subsample != (1, 1)')
+
+
+        shape = as_tensor_variable(shape)
+        broadcastable = [topgrad.type.broadcastable[0],
+                         kern.type.broadcastable[1],
+                         False, False]
+        output = kern.type.__class__(dtype=kern.type.dtype,
+                                     broadcastable=broadcastable)()
+        output = kern.type.clone(broadcastable=broadcastable)()
+        return Apply(self, [kern, topgrad, shape], [output])
+
+
+    def perform(self, node, nodename, inp, out_):
+        raise NotImplementedError('AbstractConv2d_gradWeight theano optimization failed')
+
+    def grad(self, inp, grads):
+        weights, top = inp[:2]
+        bottom, = grads
+        d_weights = AbstractConv2d_gradWeights(self.imshp, self.kshp,
+                                               self.bsize,
+                                               self.border_mode,
+                                               self.subsample)(bottom, top, weights.shape[-2:])
+        d_top = AbstractConv2d(self.imshp, self.filter_shape, self.bsize,
+                               self.border_mode, self.subsample)(bottom, weights)
+        d_height_width = (theano.gradient.DisconnectedType()(),) * 2
+        return (d_weights, d_top) + d_height_width
+
+    def connection_pattern(self, node):
+        return [[1], [1], [0], [0]]  # no connection to height, width
+
+
+### Optimizations should be move in their appropriate files
+
+### move to Gpu optimization
+### Do not replace the AbstractOpt only the inputs
+### Abstract Ops is replaced layer by device_specialized opt
+@local_optimizer([gpu_from_host, AbstractConv2d,
+                  AbstractConv2d_gradWeights,
+                  AbstractConv2d_gradInputs])
+def local_conv2d_gpu_conv(node):
+    """
+    gpu_from_host(AbstractConv) -> AbstractConv(gpu_from_host)
+
+    AbstractConv(host_from_gpu) -> host_from_gpu(AbstractConv)
+    """
+    if isinstance(node.op, GpuFromHost):
+        #gpu_from_host(conv) -> gpu_conv(gpu_from_host)
+        host_input = node.inputs[0]
+        if host_input.owner and \
+                (isinstance(host_input.owner.op, AbstractConv2d) or
+                 isinstance(host_input.owner.op, AbstractConv2d_gradWeights) or
+                 isinstance(host_input.owner.op, AbstractConv2d_gradInputs)):
+
+            conv = host_input.owner.op
+            inps = list(host_input.owner.inputs)
+            inps[0] = gpu_from_host(inps[0])
+            inps[1] = gpu_from_host(inps[1])
```

Comment:
Check `as_cudandarray_variable` instead of `gpu_from_host`, as one of the inputs could already be on the GPU.